### PR TITLE
SONAR-8795 Allow searching by project key and name in api/components/search_projects

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.StreamSupport;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.core.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class FilterParser {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SearchProjectsAction.java [hunk 2]
```diff
@@ -207,8 +217,10 @@ private SearchResults searchData(DbSession dbSession, SearchProjectsRequest requ
   }
 
   @CheckForNull
-  private Set<String> buildFilterOnFavoriteProjectUuids(List<String> criteria, Set<String> favoriteProjectUuids) {
-    if (hasIsFavoriteCriterion(criteria)) {
+  private Set<String> buildFilterOnFavoriteProjectUuids(List<FilterParser.Criterion> criteria, Set<String> favoriteProjectUuids) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SearchProjectsAction.java]: MAJOR Make "buildFilterOnFavoriteProjectUuids" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [NewIndex.java#L277](https://github.com/SonarSource/sonarqube/blob/145f62ab424b2742d42c5d106ca1cb9968c908ea/server/sonar-server/src/main/java/org/sonar/server/es/NewIndex.java#L277): Define a constant instead of duplicating this literal "index" 3 times. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [NewIndex.java#L277](https://github.com/SonarSource/sonarqube/blob/145f62ab424b2742d42c5d106ca1cb9968c908ea/server/sonar-server/src/main/java/org/sonar/server/es/NewIndex.java#L277): Define a constant instead of duplicating this literal "not_analyzed" 3 times. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProjectMeasuresIndex.java#L203](https://github.com/SonarSource/sonarqube/blob/145f62ab424b2742d42c5d106ca1cb9968c908ea/server/sonar-server/src/main/java/org/sonar/server/measure/index/ProjectMeasuresIndex.java#L203): Define a constant instead of duplicating this literal "nested_" 3 times. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [textsearch](https://github.com/SonarSource/sonarqube/blob/145f62ab424b2742d42c5d106ca1cb9968c908ea/server/sonar-server/src/main/java/org/sonar/server/es/textsearch): Add a 'package-info.java' file to document the 'textsearch' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java [hunk 3]
```diff
@@ -79,7 +81,7 @@ private static Criterion tryParsingCriterionHavingValue(String criterion) {
     Criterion.Builder builder = new Criterion.Builder();
     builder.setKey(matcher.group(1));
     builder.setOperator(matcher.group(2));
-    builder.setValue(matcher.group(3));
+    builder.setValue(sanitizeValue(matcher.group(3)));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: we should use REGEXP for that, no sanitizeValue

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java [hunk 4]
```diff
@@ -39,6 +39,7 @@
 class ProjectMeasuresQueryFactory {
 
   public static final String IS_FAVORITE_CRITERION = "isFavorite";
+  public static final String QUERY_KEY = "q";
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: I prefer query that q:

- more explicit
- might confuse with another q parameter that means something else

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java [hunk 5]
```diff
@@ -65,6 +66,11 @@ private static void processCriterion(FilterParser.Criterion criterion, ProjectMe
       return;
     }
 
+    if (QUERY_KEY.equalsIgnoreCase(key)) {
+      processQuery(criterion, query);
+      return;
+    }
+
     String value = criterion.getValue();
     checkArgument(value != null, "Value cannot be null for '%s'", key);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: remark: I'm not sure about the key naming. Good for operator and values but key?

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java [hunk 6]
```diff
@@ -89,6 +95,18 @@ private static void processLanguages(FilterParser.Criterion criterion, ProjectMe
     }
   }
 
+  private static void processQuery(FilterParser.Criterion criterion, ProjectMeasuresQuery query) {
+    String operatorValue = criterion.getOperator();
+    String value = criterion.getValue();
+    if (value == null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: checkArgument?

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java [hunk 7]
```diff
@@ -89,6 +95,18 @@ private static void processLanguages(FilterParser.Criterion criterion, ProjectMe
...
+      throw new IllegalArgumentException("Query is invalid");
+    }
+    if (!EQ.getValue().equalsIgnoreCase(operatorValue)) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: checkArgument?

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java [hunk 8]
```diff
@@ -34,10 +34,12 @@
 
 public class FilterParser {
 
+  private static final String QUOTE = "'";
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: we should handle double quote

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: I thought we have decided yesterday to keep usage of q ?

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: What do you propose ?

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java (hunk 8 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: I prefer to use only one character, it allows to use double quote in the search (note that we could do the reverse : use double only quotes)

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: After discussion, only double quote should handled for the moment

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java [hunk 9]
```diff
@@ -34,10 +34,12 @@
...
+
   private static final Splitter CRITERIA_SPLITTER = Splitter.on(Pattern.compile("and", Pattern.CASE_INSENSITIVE));
   private static final Splitter IN_VALUES_SPLITTER = Splitter.on(",").omitEmptyStrings().trimResults();
 
-  private static final Pattern PATTERN_HAVING_VALUE = Pattern.compile("(\\w+)\\s+(\\S+)\\s+(\\w+)");
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/FilterParser.java]: "(\\w+)\\s+(\\S+)\\s+(\"?\\w+\"?)"

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/ProjectMeasuresQueryFactory.java]: Let's go for query

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [textsearch](https://github.com/SonarSource/sonarqube/blob/79b9eaad5a1a90c97add827fdd6df129a44d9662/server/sonar-server/src/main/java/org/sonar/server/es/textsearch): Add a 'package-info.java' file to document the 'textsearch' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**